import os
import sys
from pathlib import Path


refs = Path("references")
dbs = Path("dbs")
tax = Path("taxonomy")


db16S = multiext(f"{dbs}/ncbi16S/1.1/bacteria_and_archea.16SrRNA", ".fna", ".fna.nhr", ".fna.nin", ".fna.nog", ".fna.nsd", ".fna.nsi", ".fna.nsq", "_id_and_taxonomy.txt")
adapters =  refs / "synthetic" / "stephenturner" / "adapters_combined_256_unique.fasta"

#metaphlan = multiext(bacteria_and_archea.16SrRNA.fna)


default_container = "docker://ghcr.io/vdblab/utility:0b"

rule all:
    input:
        db16S,
        adapters,

rule install_taxonkit_taxonomy:
    """ I wish this wasn't needed, but taxonkit via singularity doesn't seem
    to be able to access where taxonkit installs the nodes db
    Further, taxonkit needs its OWN copy, they can't just
     use the unpacked taxdump.  Other tools might use the taxdump (eg kraken)
    so we leave it here
    """
    container: default_container
    output:
        dir= directory(tax / "NCBI"),
        tkdir= directory(tax / "NCBI" / "taxonkit" ),
        nodes = tax / "NCBI" / "nodes.dmp",
        tknodes = tax / "NCBI" / "taxonkit" / "nodes.dmp",

    shell:"""
    cd {output.dir}
    wget -c ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz
    tar -zxvf taxdump.tar.gz
    cp  names.dmp nodes.dmp delnodes.dmp merged.dmp taxonkit
    rm taxdump.tar.gz
    """

rule ncbi_16S:
    """
    Where do the 16S database files come from?

    As far as I can tell, it is downloaded from <https://ftp.ncbi.nlm.nih.gov/blast/db/16S_ribosomal_RNA.tar.gz>

    that is then post-processed to get the labels file using `blastdbcmd` (from the blast commandline package) to get the accessions and taxids

    Then, because `blastdbcmd` doesn't give the whole taxonomy hierarchy, use taxonkit to extract the data in a way that works with the database
    """
    container:
        "docker://nickp60/micro16s-blast",
    input:
        taxdb_nodes = "taxonomy/NCBI/nodes.dmp",
    output:
        # it doens't work to just give db16S as the output; complains about a missing comma?
        multiext(f"{dbs}/ncbi16S/1.1/bacteria_and_archea.16SrRNA", ".fna", ".fna.nhr", ".fna.nin", ".fna.nog", ".fna.nsd", ".fna.nsi", ".fna.nsq", "_id_and_taxonomy.txt")
    params:
        taxdir=tax / "NCBI" / "taxonkit",
    shell:"""
    export TAXONKIT_DB=${{PWD}}/{params.taxdir}/
    ls $TAXONKIT_DB
    cd dbs/ncbi16S/1.1/
    wget https://ftp.ncbi.nlm.nih.gov/blast/db/16S_ribosomal_RNA.tar.gz && tar xzf 16S_ribosomal_RNA.tar.gz
    blastdbcmd -db 16S_ribosomal_RNA -outfmt "%a %T" -entry "all" > acc_taxid
    cut -f 2 -d" "  acc_taxid  | taxonkit lineage  | awk '$2!=""'  | taxonkit reformat -P | cut -f 3 | paste acc_taxid - > bacteria_and_archea.16SrRNA.id_and_taxonomy.txt
    """

rule adapters:
    container: default_container
    output:
        adapters,
    # lock this to a specific hash
    shell:""" wget  https://raw.githubusercontent.com/stephenturner/adapters/master/adapters_combined_256_unique.fasta -O {output}
    """



rule get_hg38:
    container: default_container
    output:
        refs / "human" / "hg38" / "hg38.fa.masked.gz"
    shell:"""
    wget 'ftp://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.masked.gz' -O {output}
    """
rule get_GRCm39:
    output:
         refs / "mouse" / "GRCm39" / "GCA_000001635.9_GRCm39_genomic.fna.gz"
    shell:"""
    wget https://ftp.ncbi.nlm.nih.gov/genomes/genbank/vertebrate_mammalian/Mus_musculus/latest_assembly_versions/GCA_000001635.9_GRCm39/GCA_000001635.9_GRCm39_genomic.fna.gz -O {output}
    """

rule minimap:
    """
    we use Minimap to check for host contamination.
    """
    input:
        ref=f"{refs}/{{org}}/{{name}}/{{thisref}}.gz",
    output:
        ref=f"{refs}/{{org}}/{{name}}/{{thisref}}.gz.mmi",
    container:
        "docker://evolbioinfo/minimap2:v2.24"
    resources:
        mem_mb=64 * 1024,
    threads: 16
    shell:" minimap2  -x sr --sr -u both --secondary=no -N 30 -c -d {input.ref}.gz.mmi {input.ref}.gz "



#rule metaphlan:
#    shell:
